Allow for configurable SSH username

ebauman opened the pull request:
Signed-off-by: Eamon Bauman <eamon@eamonbauman.com>

**What this PR does / why we need it**: Currently the ssh username is hardcoded to "ubuntu". This fixes that. 

**Which issue(s) this PR fixes**: fixes https://github.com/hobbyfarm/hobbyfarm/issues/83

ebauman commented:
This additionally retains the default functionality via "defaultUsername" constant, which is set to Ubuntu. That way, environments that do not set the ssh username for a template will not be impacted by this change.

Oats87 commented on pkg/controllers/vmsetcontroller/vmsetcontroller.go:
```diff
@@ -225,6 +226,10 @@ func (v *VirtualMachineSetController) reconcileVirtualMachineSet(vmset *hfv1.Vir
 		glog.V(5).Infof("provisioning %d vms", needed)
 		// this code is so... verbose...
 		for i := 0; i < needed; i++ {
+			sshUser, exists := env.Spec.TemplateMapping[vmt.Name]["ssh_username"]
```
This change will only set the sshUser when a VMSet is created, which breaks existing behavior. I think this type of behavior should be moved/duplicated to a level closer to the VM, so that existing VMs are not suddenly broken.

ebauman replied:
Would those existing VMs be broken if the default username is still "ubuntu"?

Oats87 replied:
Theoretically no, if you were able to set the existing VMs (or default in the code somewhere) then they would be fine. I am also thinking of use cases where VMs are created outside of using a VMset, like dynamic provisioning and so forth

ebauman replied:
Makes sense. The SshUsername will need to be set on the VM object whenever a VM itself is created, which in current config is either via the vmset controller or the dynamic bind controller. Does it make sense to add this functionality to the dynamic bind controller and leave this here? Or else how would one go about setting the username on the VM resource? Set it post-creation via an independent VM controller of some kind?

Oats87 replied:
Honestly, I'm not opposed to behavior where it just defaults to `ubuntu` and if you want anything else you need to explicitly set it yourself

ebauman replied:
I think that's what we have now (except I need to implement that in the DBC, my oversight). 
Or are you saying don't bother with this bit here in the vmset controller _at all_, and let the shell handle the default vs. vmspec behavior?

Oats87 replied:
Having this functionality both here and in the DBC would for sure work. 

The shell should be able to handle the "catch-all" cases i.e. where a VM template does not have a defined ssh username, a VM is created out of band of VMSet (by some other unknown controller), or a VM was created pre-definition of ssh_username

ebauman replied:
Ok. I will add this to the DBC and push for your review!

Oats87 commented on pkg/shell/shell.go:
```diff
@@ -116,7 +116,7 @@ func (sp ShellProxy) ConnectFunc(w http.ResponseWriter, r *http.Request) {
 	// now use the secret and ssh off to something
 	config := &ssh.ClientConfig{
-		User: "ubuntu",
+		User: vm.Spec.SshUsername,
```
Maybe the default logic should also be put here, that way no VM is broken accidentally

Oats87 commented on pkg/apis/hobbyfarm.io/v1/types.go:
```diff
@@ -38,6 +38,7 @@ type VirtualMachineList struct {
 type VirtualMachineSpec struct {
 	Id                       string `json:"id"`
 	VirtualMachineTemplateId string `json:"vm_template_id"`
+	SshUsername				 string `json:"ssh_username"`
```
Formatting here does not appear to be correct, tabs v spaces

ebauman commented:
I've done the following after your initial review @Oats87:
- Cleaned up some tabs vs spaces
- Added default username behavior to the shell server
- Implemented the ssh username functionality in the dynamic bind controller
- Standardized constant names

Oats87 reviewed (changes requested):
Looking at this, I'm not actually sure we want to default setting the `ssh_username` in the dynamic bind controller or vmset controller, because it will eventually get set by the shell server regardless. There's a lot of "default" logic here and three different definitions of the `ubuntu` constant

ebauman commented:
Yeah, smart. Okay, I've changed the controllers to only set the ssh user field if that is present, otherwise the field does not get set at all. Then the default behavior can be taken care of in the shell server.

Oats87 commented on pkg/shell/shell.go:
```diff
@@ -114,9 +115,14 @@ func (sp ShellProxy) ConnectFunc(w http.ResponseWriter, r *http.Request) {
 	}
 
+	sshUsername := vm.Spec.SshUsername
+	if len(sshUsername) < 1 {
```
Just curious, why are you doing `< 1` instead of `== 0`?

ebauman replied:
Just what my mind did in deference to doing `> 0` in other checks lol.

Oats87 reviewed (approved):
lgtm
